### ISSUE TITLE
losing writes during failover

### ISSUE THREAD

**Issue by josephnatale** [aterleto/jedis#6]
## Setup

* 2 clusters running locally
  * cluster 1  - locahost:6379
  * cluster 2 - localhost:6380
* the test application is just a unit test looping doing 1 write / second 
* to create the failure scenario I'm simply shutting down the primary cluster mid test

## results

The failover works as expected but I do see some lost writes when the library flips between clusters, this is behavior I can consistently recreate

### logs from unit test
```
2023-05-25 10:25:02.777  INFO 51746 --- [    Test worker] c.d.p.h.r.config.JedisFunctionalTests    : Started JedisFunctionalTests in 0.741 seconds (JVM running for 1.551)
writing value 0
writing value 1
writing value 2
writing value 3
writing value 4
writing value 5
writing value 6
writing value 7
writing value 8
2023-05-25 10:25:11.152  WARN 51746 --- [    Test worker] j.p.MultiClusterPooledConnectionProvider : 2023-05-25T10:25:11.152388-05:00[America/Chicago]: Retry 'cluster:1:127.0.0.1:6379', waiting PT0.5S until attempt '1'. Last attempt failed with exception 'redis.clients.jedis.exceptions.JedisConnectionException: Unexpected end of stream.'.
2023-05-25 10:25:11.663  WARN 51746 --- [    Test worker] j.p.MultiClusterPooledConnectionProvider : 2023-05-25T10:25:11.663353-05:00[America/Chicago]: Retry 'cluster:1:127.0.0.1:6379', waiting PT1S until attempt '2'. Last attempt failed with exception 'redis.clients.jedis.exceptions.JedisConnectionException: Failed to connect to any host resolved for DNS name.'.
2023-05-25 10:25:12.670 ERROR 51746 --- [    Test worker] j.p.MultiClusterPooledConnectionProvider : 2023-05-25T10:25:12.669767-05:00[America/Chicago]: Retry 'cluster:1:127.0.0.1:6379' recorded a failed retry attempt. Number of retry attempts: '3'. Giving up. Last exception was: 'redis.clients.jedis.exceptions.JedisConnectionException: Failed to connect to any host resolved for DNS name.'.
2023-05-25 10:25:12.670 ERROR 51746 --- [    Test worker] j.p.MultiClusterPooledConnectionProvider : 2023-05-25T10:25:12.670730-05:00[America/Chicago]: CircuitBreaker 'cluster:1:127.0.0.1:6379' recorded an error: 'redis.clients.jedis.exceptions.JedisConnectionException: Failed to connect to any host resolved for DNS name.'. Elapsed time: 1535 ms
catching connection exception
writing value 9
2023-05-25 10:25:13.675  WARN 51746 --- [    Test worker] j.p.MultiClusterPooledConnectionProvider : 2023-05-25T10:25:13.675025-05:00[America/Chicago]: Retry 'cluster:1:127.0.0.1:6379', waiting PT0.5S until attempt '1'. Last attempt failed with exception 'redis.clients.jedis.exceptions.JedisConnectionException: Failed to connect to any host resolved for DNS name.'.
2023-05-25 10:25:14.178  WARN 51746 --- [    Test worker] j.p.MultiClusterPooledConnectionProvider : 2023-05-25T10:25:14.178759-05:00[America/Chicago]: Retry 'cluster:1:127.0.0.1:6379', waiting PT1S until attempt '2'. Last attempt failed with exception 'redis.clients.jedis.exceptions.JedisConnectionException: Failed to connect to any host resolved for DNS name.'.
2023-05-25 10:25:15.185 ERROR 51746 --- [    Test worker] j.p.MultiClusterPooledConnectionProvider : 2023-05-25T10:25:15.185271-05:00[America/Chicago]: Retry 'cluster:1:127.0.0.1:6379' recorded a failed retry attempt. Number of retry attempts: '3'. Giving up. Last exception was: 'redis.clients.jedis.exceptions.JedisConnectionException: Failed to connect to any host resolved for DNS name.'.
2023-05-25 10:25:15.185 ERROR 51746 --- [    Test worker] j.p.MultiClusterPooledConnectionProvider : 2023-05-25T10:25:15.185748-05:00[America/Chicago]: CircuitBreaker 'cluster:1:127.0.0.1:6379' recorded an error: 'redis.clients.jedis.exceptions.JedisConnectionException: Failed to connect to any host resolved for DNS name.'. Elapsed time: 1512 ms
2023-05-25 10:25:15.186 ERROR 51746 --- [    Test worker] j.p.MultiClusterPooledConnectionProvider : 2023-05-25T10:25:15.186210-05:00[America/Chicago]: CircuitBreaker 'cluster:1:127.0.0.1:6379' exceeded failure rate threshold. Current failure rate: 20.0
2023-05-25 10:25:15.190  WARN 51746 --- [    Test worker] j.p.MultiClusterPooledConnectionProvider : 2023-05-25T10:25:15.190612-05:00[America/Chicago]: CircuitBreaker 'cluster:1:127.0.0.1:6379' changed state from CLOSED to OPEN
catching connection exception
writing value 10
2023-05-25 10:25:16.195 ERROR 51746 --- [    Test worker] j.p.MultiClusterPooledConnectionProvider : 2023-05-25T10:25:16.195451-05:00[America/Chicago]: CircuitBreaker 'cluster:1:127.0.0.1:6379' recorded a call which was not permitted.
2023-05-25 10:25:16.196  WARN 51746 --- [    Test worker] j.p.MultiClusterPooledConnectionProvider : 2023-05-25T10:25:16.196854-05:00[America/Chicago]: CircuitBreaker 'cluster:1:127.0.0.1:6379' changed state from OPEN to FORCED_OPEN
2023-05-25 10:25:16.197  WARN 51746 --- [    Test worker] j.p.MultiClusterPooledConnectionProvider : Cluster/database endpoint successfully updated from 'cluster:1:127.0.0.1:6379' to 'cluster:2:127.0.0.1:6380'
opening circuit on cluster cluster:2:127.0.0.1:6380
writing value 11
writing value 12
writing value 13
writing value 14
writing value 15
writing value 16
writing value 17
writing value 18
writing value 19
BUILD SUCCESSFUL in 25s
```

### cluster 1 keys
```
jnatale@MLRW20147716 (ocp:apps-dev-1:h2-dev) ~
$ redis-cli
127.0.0.1:6379> keys *
1) "0"
2) "3"
3) "6"
4) "4"
5) "7"
6) "1"
7) "2"
8) "5"
127.0.0.1:6379> exit
```

### cluster 2 keys 
```
jnatale@MLRW20147716 (ocp:apps-dev-1:h2-dev) ~
$ redis-cli -h localhost -p 6380
localhost:6380> keys *
 1) "12"
 2) "18"
 3) "17"
 4) "13"
 5) "11"
 6) "10"
 7) "19"
 8) "14"
 9) "15"
10) "16"
localhost:6380> 
```

**Keys 8 and 9 are missing!!**

## Initial thoughts

* Its unclear at this point how many writes we might lose in a high volume / multithreaded environment
* Even with some writes being lost we are still incredibly better off than if we didn't have the capability all together (suppose this might depend on your use case)
* Might be worth a backlog item to look into this further
* Might also want to be sure to document the behavior so consumers know what to expect

**Comment by aterleto** [comment]
Hi Joe, this is related to [Issue#7](https://github.com/aterleto/jedis/issues/7). 

Based on the way you wrote the test, it is expected behavior that you lost value 8 and 9. This is because you likely have a minCalls set to 2 which means that within a sliding window you can only incur 2 failed commands before the circuit breaker opens on the next subsequent command. If you want to avoid losing data, try adding retry logic into your application. 

Let's use your existing test as an example. First you sent value 8, which failed. Then you sent value 9, which failed. And then you sent value 10, which succeeded. So we know that after the 2nd failure your test successfully failed-over. Therefore let's add retry logic into your application that attempts 2 retries. The expected outcome would be a successful commit of value 8 on the 2nd retry in which case you would not have lost data.

Now you might be saying that 2 or more retires is too much. Well that might be true however keep in mind that in a production environment these commands will very likely be concurrent (not sequential like your test). So even if you had retried just once, in a production environment with at least 3 concurrent ops/s, the circuit breaker would have opened before the first retry was attempted and therefore your command would have been successful.

Hope that makes sense and gives you some ideas on how you can tackle your unique needs.

**Comment by josephnatale** [comment]
closing per #7

**Comment by aterleto** [comment]
Good Morning Joe!

Your presumption was totally logical. Z

However, the way resilience4j works is that retries do not count against the circuit breaker statistics because it wants to avoid jittery network behavior that will likely stabilize in a matter of milliseconds therefore avoiding a false positive. 

I considering wrapping another layer of retry logic around the resilience4j built-in capability in order to solve for what you intended but that seemed like it might become confusing to the end-user and therefore I decided to leave it out. It can still be added of course using the very same resilience4j framework, another open-source framework (i.e. failsafe), or manually.

In fact, since you are already adding the resilience4j dependencies to your application you might want to take advantage of resilience4j for this very purpose by creating a retry decorator in the same way I did. I can point you to the class name if you are interested.
